File: geocoder-openstreetmap/src/main/java/com/prolificinteractive/geocoder/api/openstreetmaps/OSMAddress.java
package com.prolificinteractive.geocoder.api.openstreetmaps;

import android.support.annotation.Nullable;
import com.google.gson.annotations.SerializedName;
import com.prolificinteractive.geocoder.model.Address;
import com.prolificinteractive.geocoder.model.Location;
import java.util.List;

class OSMAddress {

  @SerializedName("boundingbox")
  @Nullable public List<String> boundingbox;

  @SerializedName("address")
  @Nullable public OSMAddressDetail address;

  @SerializedName("lon")
  @Nullable public String lon;

  @SerializedName("lat")
  @Nullable public String lat;

  @SerializedName("display_name")
  @Nullable public String displayName;

  @SerializedName("type")
  @Nullable public String type;

  @Nullable public Address getAddress() {
    return Address.builder()
        .country(this.address.country)
        .administrativeAreaLevel2(this.address.city)
        .administrativeAreaLevel1(this.address.state)
        .neighborhood(this.address.neighbourhood)
        .postalCode(this.address.postcode)
        .formattedAddress(this.displayName)
        .location(Location.builder()
            .longitude(Double.parseDouble(this.lon))
            .latitude(Double.parseDouble(this.lat))
            .build())
        .build();
  }
}

File: geocoder-openstreetmap/src/main/java/com/prolificinteractive/geocoder/api/openstreetmaps/OSMAddress.kt
package com.prolificinteractive.geocoder.api.openstreetmaps

import com.google.gson.annotations.SerializedName
import com.prolificinteractive.geocoder.model.Address
import com.prolificinteractive.geocoder.model.Location

data class OSMAddress(

    @SerializedName("boundingbox")
    val boundingbox: List<String>? = null,

    @SerializedName("address")
    val osmAddress: OSMAddressDetail? = null,

    @SerializedName("lon")
    val lon: String? = null,

    @SerializedName("lat")
    val lat: String? = null,

    @SerializedName("display_name")
    val displayName: String? = null,

    @SerializedName("type")
    val type: String? = null
) {
  fun getAddress(): Address {
    return Address(
        country = this.osmAddress?.country,
        administrativeAreaLevel2 = this.osmAddress?.city,
        administrativeAreaLevel1 = this.osmAddress?.state,
        neighborhood = this.osmAddress?.neighbourhood,
        postalCode = this.osmAddress?.postcode,
        formattedAddress = this.displayName,
        location = Location(
            longitude = this.lon?.toDouble(),
            latitude = this.lat?.toDouble()
        )
    )
  }
}